controllers/nstemplateset/cluster_resources.go

```diff
@@ -24,157 +19,131 @@ type clusterResourcesManager struct {
-		if err != nil {
-			return false, r.wrapErrorWithStatusUpdateForClusterResourceFailure(gvkCtx, nsTmplSet, err,
-				"failed to list existing cluster resources of GVK '%v'", clusterResourceKind.gvk)
+	for _, newObj := range newObjs {
+		if !shouldCreate(newObj, nsTmplSet) {
+			continue
 		}
 
-		// if there are more than one existing, then check if there is any that should be updated or deleted
-		if len(currentObjects) > 0 {
-			updatedOrDeleted, err := r.updateOrDeleteRedundant(gvkCtx, currentObjects, newObjs, tierTemplate, nsTmplSet)
-			if err != nil {
-				return false, r.wrapErrorWithStatusUpdate(gvkCtx, nsTmplSet, r.setStatusUpdateFailed,
-					err, "failed to update/delete existing cluster resources of GVK '%v'", clusterResourceKind.gvk)
-			}
-			if updatedOrDeleted {
-				return true, err
+		// by removing the new objects from the current objects, we are going to be left with the objects
+		// that should be removed from the cluster after we're done with this loop
+		for oldIdx, curObj := range currentObjects {
+			if curObj.GetName() == newObj.GetName() && curObj.GetNamespace() == newObj.GetNamespace() && curObj.GetObjectKind().GroupVersionKind() == newObj.GetObjectKind().GroupVersionKind() {
+				currentObjects = slices.Delete(currentObjects, oldIdx, oldIdx+1)
+				break
 			}
 		}
-		// if none was found to be either updated or deleted or if there is no existing object available,
-		// then check if there is any object to be created
-		if len(newObjs) > 0 {
-			anyCreated, err := r.createMissing(gvkCtx, currentObjects, newObjs, tierTemplate, nsTmplSet)
-			if err != nil {
-				return false, r.wrapErrorWithStatusUpdate(gvkCtx, nsTmplSet, r.setStatusClusterResourcesProvisionFailed,
-					err, "failed to create missing cluster resource of GVK '%v'", clusterResourceKind.gvk)
-			}
-			if anyCreated {
-				return true, nil
+
+		if err := changeStatusIfNeeded(); err != nil {
+			return err
+		}
+
+		// create or update the resource
+		if _, err = r.apply(ctx, nsTmplSet, newTierTemplate, newObj); err != nil {
+			err := fmt.Errorf("failed to apply changes to the cluster resource %s, %s: %w", newObj.GetName(), newObj.GetObjectKind().GroupVersionKind().String(), err)
+			return r.wrapErrorWithStatusUpdate(ctx, nsTmplSet, failureStatusReason, err, "failure while syncing cluster resources")
+		}
+	}
+
+	// what we're left with here is the list of currently existing objects that are no longer present in the template.
+	// we need to delete them
+	for _, obj := range currentObjects {
+		if err := changeStatusIfNeeded(); err != nil {
+			return err
+		}
+		if err := r.Client.Delete(ctx, obj); err != nil {
+			if !errors.IsNotFound(err) {
+				err := fmt.Errorf("failed to delete the cluster resource %s, %s: %w", obj.GetName(), obj.GetObjectKind().GroupVersionKind().String(), err)
+				return r.wrapErrorWithStatusUpdate(ctx, nsTmplSet, failureStatusReason, err, "failure while syncing cluster resources")
```

Comment:
have you considered using the already existing `deleteObsoleteObjects` func?

Reply:
I have and opted not to use it because it is meant to be used with 1-obj-per-reconcile approach (and is used when dealing with namespaces and spaces).

This func is meant to delete 1 obsolete resource at a time without updating the status - which is done elsewhere.

What we're doing here is the full procedure for all resources - create new, update existing, delete obsolete and make updates to the status as appropriate. I suspect that this whole procedure will be reusable in the other scenarios once we get to them, but I didn't want to mix functions used in the 1-obj-per-reconcile approach with the all-at-once approach.